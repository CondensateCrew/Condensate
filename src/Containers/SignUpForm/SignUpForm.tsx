import React, { useState, useEffect } from 'react';
import { UserSignupPosting } from '../../interfaces';
import InputElement from '../../Components/InputElement/InputElement';
import { validateCredentials } from '../../_utils';
import { postUser } from 'apiCalls/apiCalls';
import LoadingImage from 'Components/LoadingImage/LoadingImage';

type checkedInputType = 'first-name' | 'last-name' | 'email' | 'password' | 'repeat-password';

interface Props {
  isLogin: boolean,
  toggleTab: (login: boolean) => void,
  setCookie: (name: string, value: any, options?: object) => void
};

interface UserPosted {
  first_name: string,
  last_name: string,
  password: string,
  email: string,
}

const SignUpForm: React.FC<Props> = ({ isLogin, toggleTab, setCookie }) => {
  const [ user, setUser ] = useState<UserSignupPosting>({
    email: '',
    password: '',
    firstName: '',
    lastName: '',
  });
  const [ disabled, setDisabled ] = useState<boolean>(true);
  const [ repeatPassword, setPassword ] = useState<string>('');
  const [ error, setError ] = useState<string>('');
  const [ isLoading, setIsLoading ] = useState<boolean>(false);

  const validateButton = (): void => {
    if (user.email === '') return setDisabled(true);
    if (user.password === '') return setDisabled(true);
    if (user.firstName === '') return setDisabled(true);
    if (user.lastName === '') return setDisabled(true);
    if (user.password !== repeatPassword) return setDisabled(true);
    setError('');
    return setDisabled(false);
  }

  const validateInputs = (): void => {
    if (user.email === '') return setError('Please, add you email');
    if (user.password === '') return setError('Please, add you password');
    if (user.firstName === '') return setError('Please, add you first name');
    if (user.lastName === '') return setError('Please, add you last name');
    if (user.password !== repeatPassword) return setError('Passwords are not the same');
    if (!validateCredentials(user.email)) return setError('Please, enter a valid email');
    return setError('');
  };

  useEffect(validateButton, [ user, repeatPassword ]);

  const submitUser = async (event: React.MouseEvent) => {
    event.preventDefault();
    validateInputs();

    if (error === '') await postNewUser();
  }

  const postNewUser = async () => {
    try {
      const { firstName, lastName, password, email } = user;
      setDisabled(true);
      setIsLoading(true);
      const userNew: UserPosted = {
        first_name: firstName,
        last_name: lastName,
        password: password,
        email: email,
      };
      const { token, first_name, last_name} = await postUser(userNew);
      setCookie('token', token, { path: '/' });
      setCookie('name', `${first_name}-${last_name}`, { path: '/' });
    }
    catch (error) {
      setError(error.message);
      setIsLoading(false);
    }
  }

  const toggleForm = (): void => toggleTab(!isLogin);

  const inputs: checkedInputType[] = ['first-name', 'last-name', 'email', 'password', 'repeat-password'];
  const inputsElements: JSX.Element[] = inputs.map((input: checkedInputType) => (
    <InputElement
      key={input} typeInput={input}
      {... { isLoading, user, repeatPassword, setPassword, setUser, setError }}
    />
  ));

  return (
    <div className="sign-up-form">
      <header>
        <h2 className="active">Sign Up</h2>
        <h2 className="hidden" onClick={toggleForm}>Login</h2>
      </header>
      <form>
        {error !== '' && <p className="error-notification">{error}</p>}
        {inputsElements}
        <button disabled={disabled} onClick={submitUser}>
        {
          isLoading
            ? <LoadingImage />
            : 'Sign Up'
        }
        </button>
      </form>
    </div>
  );
}

export default SignUpForm;
