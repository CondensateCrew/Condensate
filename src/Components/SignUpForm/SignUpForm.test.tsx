/**
 * @jest-environment jsdom
 */

import React from 'react';
import { shallow } from 'enzyme';
import SignUpForm from './SignUpForm';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useDispatch: () => mockDispatch
}));

describe('SignUpForm component', () => {
  const mockToggle = jest.fn();
  const mockSetIsLoading = jest.fn();
  const mockSetCookie = jest.fn();

  const mockProps = {
    isLoading: false,
    isLogin: false,
    toggleTab: mockToggle,
    setIsLoading: mockSetIsLoading,
    setCookie: mockSetCookie
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    const wrapper: any = shallow(<SignUpForm {...mockProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  // describe('Change Events', () => {
  //   it('should render a disabled button if not all form fields are filled', () => {
  //     const mockEvent: Object = { target: { value: 'Goose' } }
  //     const wrapper: any = shallow(<SignUpForm {...mockProps} />);
  //
  //     wrapper.find('input').first().simulate('change', mockEvent);
  //     expect(wrapper.find('button').getDOMNode().disabled).toEqual(true);
  //   });
  //
  //   it('should render an enabled button if all form fields are filled', () => {
  //       const mockEvent: Object = { target: { value: 'Goose' } };
  //       const mockEvent2: Object = { target: { value: 'Red' } };
  //       const mockEvent3: Object = { target: { value: 'loosegoose@email.com' } };
  //       const mockEvent4: Object = { target: { value: 'password123' } };
  //       const mockEvent5: Object = { target: { value: 'password123' } };
  //
  //       const wrapper: any = shallow(<SignUpForm {...mockProps} />);
  //
  //       wrapper.find('input').first().simulate('change', mockEvent);
  //       wrapper.find('#last-name').first().simulate('change', mockEvent2);
  //       wrapper.find('#email').first().simulate('change', mockEvent3);
  //       wrapper.find('#password').first().simulate('change', mockEvent4);
  //       wrapper.find('input').last().simulate('change', mockEvent5);
  //
  //       expect(wrapper.find('button').getDOMNode().disabled).toEqual(false);
  //   });
  // });
});
