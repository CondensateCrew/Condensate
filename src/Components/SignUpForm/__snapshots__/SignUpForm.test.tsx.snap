// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm component should match the snapshot 1`] = `
<div
  className="sign-up-form"
>
  <header>
    <h2
      className="hidden-tab"
      onClick={[Function]}
    >
      Login
    </h2>
    <h2
      className="active-tab"
    >
      Sign Up
    </h2>
  </header>
  <form>
    <InputElement
      isLoading={false}
      key="first-name"
      repeatPassword=""
      setError={[Function]}
      setPassword={[Function]}
      setUser={[Function]}
      typeInput="first-name"
      user={
        Object {
          "email": "",
          "firstName": "",
          "lastName": "",
          "password": "",
        }
      }
    />
    <InputElement
      isLoading={false}
      key="last-name"
      repeatPassword=""
      setError={[Function]}
      setPassword={[Function]}
      setUser={[Function]}
      typeInput="last-name"
      user={
        Object {
          "email": "",
          "firstName": "",
          "lastName": "",
          "password": "",
        }
      }
    />
    <InputElement
      isLoading={false}
      key="email"
      repeatPassword=""
      setError={[Function]}
      setPassword={[Function]}
      setUser={[Function]}
      typeInput="email"
      user={
        Object {
          "email": "",
          "firstName": "",
          "lastName": "",
          "password": "",
        }
      }
    />
    <InputElement
      isLoading={false}
      key="password"
      repeatPassword=""
      setError={[Function]}
      setPassword={[Function]}
      setUser={[Function]}
      typeInput="password"
      user={
        Object {
          "email": "",
          "firstName": "",
          "lastName": "",
          "password": "",
        }
      }
    />
    <InputElement
      isLoading={false}
      key="repeat-password"
      repeatPassword=""
      setError={[Function]}
      setPassword={[Function]}
      setUser={[Function]}
      typeInput="repeat-password"
      user={
        Object {
          "email": "",
          "firstName": "",
          "lastName": "",
          "password": "",
        }
      }
    />
    <button
      disabled={true}
      onClick={[Function]}
    >
      Sign Up
    </button>
  </form>
</div>
`;
